### PR TITLE
TimeRangeDataWidget CodeOSS Port

### PR DESCRIPTION
**Pull request by williamsyang-work**
Integrates the TimeRangeDataWidget to VSCode Trace Extension's trace explorer.

This requires changes in the theia-trace-extension repo: https://github.com/eclipse-cdt-cloud/theia-trace-extension/pull/991

Signed-off-by: Will Yang <william.yang@ericsson.com>

### CODE REVIEW

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (colin-grant-work) [inline vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:80]: Is this notification handled?

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (williamsyang-work) [reply vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx]: Leftover and removed.

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (williamsyang-work) [reply vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:9]: The signal manager can only reach within a JS GEC.  This was available for all views in the Theia IDE, but is contained inside of each webview for Code OSS.
We need to use a combination of signals and vscode-messages for the events to be heard where they are needed.  This architecture is a little convoluted but explained in the README under [Communcation Between Components](https://github.com/eclipse-cdt-cloud/vscode-trace-extension/blob/master/README.md#communication-between-components)

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (williamsyang-work) [reply vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:79]: I am not sure what this signal is or where it is handled.  I saw that every other webview that is in the trace-explorer fired this signal so I left it in mine.  @bhufmann, offhand, do you know if this is necessary or not?

Comment (colin-grant-work) [reply vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:79]: Usually a signal like this would be used to facilitate the following sequence:
 - User opens view
 - View provider creates webview
 - When webview is ready, view provider sends it data to render

If there's no lifecycle handling that requires the plugin side to interact with the webview when it's ready (but not before), then the ready signal is redundant (but doesn't do any harm).

Comment (colin-grant-work) [reply vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:79]: I understand that state restoration when the webview goes out of view is a problem for this PR. That is a potential use for this signal, as well. If the provider knows what signals to send to get the view in the right state, it can send those whenever it receives the ready signal. That way, every time the view re-mounts, the provider sends what's necessary to show the correct data in the view.

Alternatively, that could be handled on the view side using the VSCode's webview state restoration mechanisms. In that case, the webview side would save its state as the most recent, state-relevant signals it had received, and then fetch and reemit those signals on remount.

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (MatthewKhouzam) [inline vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:106]: experimenterror TS2341: Property 'experimentDataMap' is private and only accessible within class 'ReactTimeRangeDataWidget'.

This was tested in Gitpod.

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (MatthewKhouzam) [inline vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:119]: error TS2339: Property 'restoreState' does not exist on type 'ReactTimeRangeDataWidget'.

File: vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx
Comment (MatthewKhouzam) [inline vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx:43]: replace name with // TODO: move if need be?

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (MatthewKhouzam) [inline vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:35]: remove?

**Review by MatthewKhouzam — CHANGES_REQUESTED** [review]
Some minor comments. I cannot get it to build locally or on gitpod, I put the error codes in the comments. Hope it helps.

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (williamsyang-work) [reply vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:106]: This error is happening because this needs requires changes to the theia-trace-extension: https://github.com/eclipse-cdt-cloud/theia-trace-extension/pull/991

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (williamsyang-work) [reply vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:119]: This error is also linked to the theia-trace-extension changes.

**Comment by colin-grant-work** [comment]
Comments from demo:
 - [ ] Please remove `console.dir` fallback. In general plugins shouldn't log using the console, but should use the VSCode logging API.
 - [ ] Find and fix the BigInt conversion bug
 - [ ] With the addition of this widget, the status bar item that shows the selection is mostly redundant and should likely be removed. It appears to be generated here: 

https://github.com/eclipse-cdt-cloud/vscode-trace-extension/blob/f029c45811f50dc3874597a77a9019151cec365e/vscode-trace-extension/src/trace-viewer-panel/trace-viewer-webview-panel.ts#L178-L179

The open-ended status bar item creation implemented in the `handleStatusMessage` is a little bit troubling - the status bar shouldn't really be used for detailed communication with users, so it shouldn't be so open as to be used by anyone for anything.

**Comment by bhufmann** [comment]
> Comments from demo:
> 
> * [ ]  Please remove `console.dir` fallback. In general plugins shouldn't log using the console, but should use the VSCode logging API.
> * [ ]  Find and fix the BigInt conversion bug
> * [ ]  With the addition of this widget, the status bar item that shows the selection is mostly redundant and should likely be removed. It appears to be generated here:
> 
> https://github.com/eclipse-cdt-cloud/vscode-trace-extension/blob/f029c45811f50dc3874597a77a9019151cec365e/vscode-trace-extension/src/trace-viewer-panel/trace-viewer-webview-panel.ts#L178-L179
> 
> The open-ended status bar item creation implemented in the `handleStatusMessage` is a little bit troubling - the status bar shouldn't really be used for detailed communication with users, so it shouldn't be so open as to be used by anyone for anything. Addressing the status bar item is big enough that it can be a separate ticket, rather than getting into this PR.

Please note that the status bar also show the the time delta of the cursors which is important to have so that user can measure time deltas. If we decide to remove the time values from the status bar, the delta needs to be provided differently. The Time Range widget would be a candidate for that. 

Not sure if we should remove the cursor and delta values from the status bar. That information should be visible for the user all the time and the time range widget might be closed. I agree though, that we should not use the status bar lightly and log any arbitrary message there.

**Comment by colin-grant-work** [comment]
> Please note that the status bar also shows the the time delta of the cursors which is important.

Likely this should be added to the `TimeRangeDataWidget` if it's important. It looks like at the moment it only displays the absolute positions of view and selection, and not the deltas.

**Comment by williamsyang-work** [comment]
The build is failing because the theia-trace-extension related changes aren't currently available.

**Review by bhufmann — CHANGES_REQUESTED** [review]
When testing https://github.com/eclipse-cdt-cloud/theia-trace-extension/pull/991 I found the following issues with this PR:

1. When opening a trace the view range starts at 0. End time seems to be off as well (see next point)
2. The timestamp seems without offset applied
3. When Closing the last trace the values are not cleared
4. When trace server is not running the the timerange widget should not be rendered (as item properties and available views are not rendered)

The attached gif should show these issues:

time-range-widget-issues

**Review by bhufmann — CHANGES_REQUESTED** [review]
I put some review comments and I found some issues. There first one I have already reported earlier. The time range widget should not be rendered if the trace server is not running:
TimeRangeWidgetServerDown

The second problem is intermittent. Sometimes the view range is 0 and hence the other time values are incorrect. I have not found a deterministic way to reproduce this, but when it happened I switched between different panels of multiple open traces by clicking on the panel in the middle of the window. See the picture on how it looks like:
IncorrectOffset

The third one that I noticed is when the timerange window is minimized. Changing selection, for example, when it's minimized is not applied correctly when after maximizing it. I guess there is something handling the queue.

timerange-data-incorrect-queue

The last one, is when having multiple traces open and when close one trace that is not the selected one, then the time range widget is cleared It should be only cleared only if it's the last trace that is closed.

time-range-widget-cleared-closing

File: vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/time-range/vscode-time-range-data-widget.tsx:115]: So, if you serialize using JSONBig, all bigint variables are serialized in a special way so that when calling JSONBig.parse() a bigint is recreated. However, if the bigint value fits into a javascript number when calling JSONBig.stringify() it won't serialize it that special way and when calling JSON.big.parse() the corresponding variable is NOT a bigint. This can lead to problems later on.

So, this is the reason why the tsp-typescript-client has special code to account for that and the vscode-trace-extension has some special code for serializing/deserializing Experiment. See vscode-signal-converter.ts.

So, you will need take care of this when doing JSONBig.parse(),

BTW, we plan to streamline both vscode-signal-converter.ts and tsp-typescript-client for the bigint handling (not scope of this PR).

**Comment by MatthewKhouzam** [comment]
I am getting an endless loop on this, it does not populate with traces or experiments.

Tested with:

236c8d01395ebf2a682e651b94d666e5b09c5309 -> vscode
0533d18f609822b3d24c4d562745639d8eef4213 -> theia side

In VSCodium 1.75. and glorious Theia 1.39

**Comment by MatthewKhouzam** [comment]
Some issues!

1- needs to have same tsp-client-version as #991 from theia-trace-extension

Long text should turn red when it's impossible to be correct (or larger than end)
https://github.com/eclipse-cdt-cloud/vscode-trace-extension/assets/419535/4f194e01-9a9c-4c17-8916-8c23f19b9d0a

It's impossible to have an empty text box for "pasting in" a good timestamp (from excel)
https://github.com/eclipse-cdt-cloud/vscode-trace-extension/assets/419535/665370d8-d04a-4689-81f9-7474ca8d4794

When the time is invalid, it should be re-validated. 
https://github.com/eclipse-cdt-cloud/vscode-trace-extension/assets/419535/d4746d03-5569-4faa-a1b3-e096980a4d5b

**Comment by williamsyang-work** [comment]
Notes: 

>needs to have same tsp-client-version as #991 from theia-trace-extension
This is fixed in the most recent commit.

> Long text should turn red when it's impossible to be correct (or larger than end) https://github.com/eclipse-cdt-cloud/vscode-trace-extension/assets/419535/4f194e01-9a9c-4c17-8916-8c23f19b9d0a
The color of the input-box numbers wasn't originally red.  It was changed in [this commit in theia-trace-extension].(https://github.com/eclipse-cdt-cloud/theia-trace-extension/commit/e1842b7087b376befa485718c239b709099867a6)

This has been fixed by adding the variable `--trace-viewer-input-foreground` to vscode's css.


> It's impossible to have an empty text box for "pasting in" a good timestamp (from excel) https://github.com/eclipse-cdt-cloud/vscode-trace-extension/assets/419535/665370d8-d04a-4689-81f9-7474ca8d4794
[This has been fixed in the other repository, along with some other buggy behaviors.](https://github.com/eclipse-cdt-cloud/theia-trace-extension/pull/991/commits/99c3b151b21e3e6e32cc8b109ead9736d02b51ea)


> When the time is invalid, it should be re-validated. https://github.com/eclipse-cdt-cloud/vscode-trace-extension/assets/419535/d4746d03-5569-4faa-a1b3-e096980a4d5b

Data validation has always been on submit, not on change.  I think the change to the color of the numbers made the UX jarring.  I'm not going to change anything since fixing the font-color seemed to correct the experience.

File: vscode-trace-extension/src/trace-explorer/time-range/trace-explorer-time-range-data-webview-provider.ts
Comment (MatthewKhouzam) [inline vscode-trace-extension/src/trace-explorer/time-range/trace-explorer-time-range-data-webview-provider.ts:70]: Make sure this is correct

File: yarn.lock
Comment (MatthewKhouzam) [inline yarn.lock:1]: Update this... but then this is all good to go! great job!

**Review by MatthewKhouzam — APPROVED** [review]
Aproving assuming you change the yarn.lock and rebase. Needs a second review though.

**Comment by MatthewKhouzam** [comment]
Looks great, get the build working! :)

**Review by hriday-panchasara — APPROVED** [review]
Works great, thanks!

**Comment by marco-miller** [comment]
@williamsyang-work is looking into marking the still open conversation threads as Resolved or so. Once he is done with that, I'm in turn planning to merge this PR. Given that we have our two approvals already. Thanks.

**Review by PatrickTasse — APPROVED** [review]

**Pull request merged**
